### PR TITLE
[JAX] Use FP8 tolerances in FP8 tests

### PR DESCRIPTION
**Pull request by timmoon10**
We have had recent CI failures in `tests/jax/test_custom_call_compute.py::TestFP8Dot::test_grad_fp8_randint` when running on an L40 (e.g. [this recent build for #471](https://github.com/NVIDIA/TransformerEngine/actions/runs/6724992709/job/18278451411)). This is because this test requires bit-wise correct results. The GEMM kernel that executes on L40 can't achieve this level of accuracy, although it still within the expected numerical error for FP8.

This PR adds logic so that we can use dtype-specific tolerances within the `assert_allclose` helper function. Right now we only use it in JAX FP8 tests, but it could be applied to other tests. We can also consider porting this functionality to other framework tests.

### CODE REVIEW

**Comment by nouiz** [comment]
@mingxu1067 to review

**Comment by denera** [comment]
@timmoon10 Frederic just directed me to this PR when I asked about this test failure and I wanted to share a useful heuristic that automates the tolerances for tests like these.

You can define absolute tolerances as follows:
```
eps_relaxed = jax.lax.pow(jnp.finfo(dtype).eps, dtype(2./3.))
spacing_high = jnp.nextafter(dtype(reference_value), jnp.finfo(dtype).max) - dtype(reference_value)
spacing_low = dtype(reference_value) - jnp.nextafter(dtype(reference_value), jnp.finfo(dtype).min)
ulp = jax.lax.max(spacing_high, spacing_low)
abs_tol = jax.lax.max(jnp.finfo(dtype).eps, ulp)
```

This guarantees that the tests will never require a tolerance smaller than the unit of least precision (ULP).

Here's a table of the tolerances I see with this approach:
|          | **ref value** | **eps**       | **eps^(2/3)** | **ulp**       | abs tol   |
|----------|-----------|-----------|-----------|-----------|-----------|
| fp32     | 1.000e-02 | 1.192e-07 | 2.422e-05 | 9.313e-10 | 2.422e-05 |
|          | 1.000e-01 | 1.192e-07 | 2.422e-05 | 7.451e-09 | 2.422e-05 |
|          | 1.000e+00 | 1.192e-07 | 2.422e-05 | 1.192e-07 | 2.422e-05 |
|          | 1.000e+01 | 1.192e-07 | 2.422e-05 | 9.537e-07 | 2.422e-05 |
|          | 1.000e+02 | 1.192e-07 | 2.422e-05 | 7.629e-06 | 2.422e-05 |
| fp16     | 1.000e-02 | 9.766e-04 | 9.857e-03 | 7.629e-06 | 9.857e-03 |
|          | 1.000e-01 | 9.766e-04 | 9.857e-03 | 6.104e-05 | 9.857e-03 |
|          | 1.000e+00 | 9.766e-04 | 9.857e-03 | 9.766e-04 | 9.857e-03 |
|          | 1.000e+01 | 9.766e-04 | 9.857e-03 | 7.812e-03 | 9.857e-03 |
|          | 1.000e+02 | 9.766e-04 | 9.857e-03 | 6.250e-02 | 6.250e-02 |
| bf16     | 1.000e-02 | 7.812e-03 | 3.906e-02 | 6.104e-05 | 3.906e-02 |
|          | 1.000e-01 | 7.812e-03 | 3.906e-02 | 4.883e-04 | 3.906e-02 |
|          | 1.000e+00 | 7.812e-03 | 3.906e-02 | 7.812e-03 | 3.906e-02 |
|          | 1.000e+01 | 7.812e-03 | 3.906e-02 | 6.250e-02 | 6.250e-02 |
|          | 1.000e+02 | 7.812e-03 | 3.906e-02 | 5.000e-01 | 5.000e-01 |
| fp8_e4m3 | 1.000e-02 | 1.250e-01 | 2.344e-01 | 1.953e-03 | 2.344e-01 |
|          | 1.000e-01 | 1.250e-01 | 2.344e-01 | 7.812e-03 | 2.344e-01 |
|          | 1.000e+00 | 1.250e-01 | 2.344e-01 | 1.250e-01 | 2.344e-01 |
|          | 1.000e+01 | 1.250e-01 | 2.344e-01 | 1.000e+00 | 1.000e+00 |
|          | 1.000e+02 | 1.250e-01 | 2.344e-01 | 8.000e+00 | **8.000e+00** |
| fp8_e5m2 | 6.813e-02 | 2.500e-01 | 4.375e-01 | 1.562e-02 | 4.375e-01 |
|          | 4.642e-01 | 2.500e-01 | 4.375e-01 | 6.250e-02 | 4.375e-01 |
|          | 3.162e+00 | 2.500e-01 | 4.375e-01 | 5.000e-01 | 5.000e-01 |
|          | 2.154e+01 | 2.500e-01 | 4.375e-01 | 4.000e+00 | 4.000e+00 |
|          | 1.468e+02 | 2.500e-01 | 4.375e-01 | 3.200e+01 | 3.200e+01 |

Note that the `fp8_e4m3` tolerance computed for O(10^2) data is exactly the max error reported by the L40 test failure, which makes sense given that the tensor elements in the tests are all in the O(10^2) to O(10^3) range.

File: examples/jax/mnist/test_single_gpu_mnist.py [hunk 1]
```diff
@@ -280,7 +280,7 @@ def verify(actual):
         """Check If loss and accuracy match target"""
         desired_traing_loss = 0.055
         desired_traing_accuracy = 0.98
-        desired_test_loss = 0.035
+        desired_test_loss = 0.04
```
Comment (timmoon10) [inline examples/jax/mnist/test_single_gpu_mnist.py:283]: When I ran twice on an H100, I got 0.031 in one run and 0.034 in the next. Given this run-to-run variance, I think it's reasonable to loosen the target a little.

**Comment by denera** [comment]
@timmoon10 I filed #505 that should fix some CI failures due to an oversight in the recent custom partitioning regression tests. Currently being tested, will merge ASAP. Just wanted to give you a heads up in case your PR is affected.

**Review by denera — APPROVED** [review]

**Comment by timmoon10** [comment]
Pipeline 10644319
